File: webfluxasyncdemo/src/main/java/com/wikicoding/webfluxasyncdemo/controllers/ContactsController.java
package com.wikicoding.webfluxasyncdemo.controllers;


import com.wikicoding.webfluxasyncdemo.dto.AddContactRequest;
import com.wikicoding.webfluxasyncdemo.service.ContactsService;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("/api/v1/contacts")
@AllArgsConstructor
@Slf4j
public class ContactsController {
    private final ContactsService contactsService;

    @PostMapping
    public ResponseEntity<Object> createContact(@RequestBody AddContactRequest request) {
        log.warn("Controller endpoint running on thread {}", Thread.currentThread().getName());
        if (request.getName().trim().isEmpty() || request.getEmail().isEmpty())
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Request has to be valid");

        return ResponseEntity.status(HttpStatus.CREATED).body(contactsService.saveAsync(request.getName(), request.getEmail()));
    }

    @GetMapping("/{email}")
    public Mono<ResponseEntity<Object>> findByEmail(@PathVariable(name = "email") String email) {
        log.warn("Controller endpoint running on thread {}", Thread.currentThread().getName());

        if (email.trim().isEmpty())
            return Mono.just(ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Email has to be valid"));

        return contactsService.findByEmailAsync(email)
                .map(contact -> ResponseEntity.ok((Object) contact))
                .defaultIfEmpty(ResponseEntity.notFound().build());
    }
}

File: dotnet_async_crud_demo/dotnet_async_crud_demo/Controllers/ContactsController.cs
using dotnet_async_crud_demo.dtos;
using dotnet_async_crud_demo.Services;
using Microsoft.AspNetCore.Mvc;
using System.Diagnostics;

namespace dotnet_async_crud_demo.Controllers;

[ApiController]
[Route("api/v1/[controller]")]
public class ContactsController(ContactsService service, ILogger<ContactsController> logger) : ControllerBase
{
    [HttpPost]
    public async Task<IActionResult> saveContact([FromBody] AddContactRequest request)
    {
        logger.LogWarning("Controller endpoint running on thread {}", Thread.CurrentThread.ManagedThreadId);

        if (string.IsNullOrEmpty(request.name) || string.IsNullOrEmpty(request.email))
        {
            return BadRequest("Request has to be valid");
        }

        var contact = await service.SaveAsync(request.name, request.email);

        return CreatedAtAction(nameof(saveContact), contact);
    }

    [HttpGet("{email}")]
    public async Task<IActionResult> findByEmail([FromRoute(Name = "email")] string email)
    {
        var stopwatch = new Stopwatch();
        stopwatch.Start();

        logger.LogWarning("Controller endpoint running on thread {}", Thread.CurrentThread.ManagedThreadId);

        if (string.IsNullOrEmpty(email))
        {
            return BadRequest("Request has to be valid");
        }

        var contact = await service.FindByEmailAsync(email);

        if (contact is null)
        {
            return NotFound("Contact not found by email");
        }

        stopwatch.Stop();

        logger.LogWarning("Request took {}ms and now is on thread {}", stopwatch.ElapsedMilliseconds, Thread.CurrentThread.ManagedThreadId);

        return Ok(contact);
    }
}
